### PR TITLE
Map solidity EVM assembly to lines of code

### PR DESCRIPTION
**Pull request by anders-torbjornsen**
As per the request here https://github.com/compiler-explorer/compiler-explorer/issues/3520

image

This PR is a small tweak to the Solidity support which aims to map asm instructions to lines of code. Solc attributes each instruction a range of characters from the source, sometimes this range can be quite large - e.g. a whole function, or even a whole contract. In these cases I just use the first line, which seems to work ok as if an instruction is attributed to an entire function then attributing it to the function sig instead has a similar effect.

I also found the tags (jumpdests) kind of hard to follow so this PR also replaces as many of those tags as possible with human readable function names.

Suggestions etc are most welcome, JS isn't really my thing!

### CODE REVIEW

File: lib/compilers/solidity.js [hunk 1]
```diff
@@ -62,36 +66,158 @@ export class SolidityCompiler extends BaseCompiler {
...
+                                    tagCount: 0,
+                                };
+                            });
+                        }
+                        return generatedSources;
+                    };
+
+                    // there are two sets of generated sources, one for the code which deploys
+                    // the contract (i.e. the constructor) 'generated-sources', and the code
+                    // which is deployed and stored on-chain 'generated-sources-runtime'
```
Comment (RubenRBS) [inline lib/compilers/solidity.js:148]: Thanks for the comments! Super helpful

**Review by RubenRBS — APPROVED** [review]
Hi! This looks great, thank you very much!

We would appreciate if you could add a test to ensure that this keeps working. There are some exmaples on the `test` folder of similar cases that you could take a look (As I don't remember if we have soemthing in the documentation that could help, sorry!), or ask any question if needed, but the code looks pretty much ready to get merged otherwise :)

**Comment by RubenRBS** [comment]
Hi! We're getting ready to update the site tonight so I'll merge this without the requested tests as the code seems to work, but would really appreciate it if you could send a followup PR with those at some point :)

**Comment by RubenRBS** [comment]
Hi! What solc version are you using?
For 0.8, no line linking seems to be happening :/

For 0.7, this error shows up:
imagen

For versions 0.6 and below, no compilation is done because
imagen

**Comment by anders-torbjornsen** [comment]
@RubenRBS oh no! I'm just using the latest solc, 0.8.something, iirc there's a dir I can put solc binaries where compiler explorer will pick them up, /etc/compilers or something? Then I can test all of the compiler versions locally and get it sorted out

**Comment by anders-torbjornsen** [comment]
I'll work on some tests too after that

**Comment by RubenRBS** [comment]
You'll need to list the compilers in a properties file. My suggestion is to copy & paste the `etc/config/solidity.amazon.properties` file and rename it to `solidity.local.properties`. This will create a local config that will look for all the compilers in `/opt/compiler-explorer/solc-<VERSION>/solc` for each `VERSION`

**Comment by mattgodbolt** [comment]
Another possibility is to symlink the two (`ln -s solidity.amazon.properties etc/config/solidity.local.properties`).

**Comment by shekhirin** [comment]
It's probably related to `solc`, not your implementation but with optimizer turned on, mapping doesn't make any sense 😞 
image

https://godbolt.org/z/cj6xosj6E

**Comment by anders-torbjornsen** [comment]
@shekhirin yeah if you use --via-ir the sourcemaps all just point to the entire contract, I haven't been able to find a way around that one. If you use --optimize without --via-ir then it works ok though
